include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -3,428 +3,573 @@
 
 #include <nbl/builtin/hlsl/cpp_compat/basic.h>
 #include <nbl/builtin/hlsl/matrix_utils/matrix_traits.hlsl>
+#include <nbl/builtin/hlsl/matrix_utils/mul_output_t.hlsl>
 #include <nbl/builtin/hlsl/concepts.hlsl>
 #include <nbl/builtin/hlsl/spirv_intrinsics/core.hlsl>
 #include <nbl/builtin/hlsl/spirv_intrinsics/glsl.std.450.hlsl>
+#include <nbl/builtin/hlsl/ieee754.hlsl>
+#include <nbl/builtin/hlsl/concepts/core.hlsl>
+#include <nbl/builtin/hlsl/concepts/vector.hlsl>
+#include <nbl/builtin/hlsl/concepts/matrix.hlsl>
+#include <nbl/builtin/hlsl/cpp_compat/promote.hlsl>
+#ifndef __HLSL_VERSION
+#include <bitset>
+#endif
 
 namespace nbl
 {
 namespace hlsl
 {
 namespace cpp_compat_intrinsics_impl
 {
-template<typename T>
-struct dot_helper
-{
-	using scalar_type = typename vector_traits<T>::scalar_type;
 
-	static inline scalar_type dot_product(NBL_CONST_REF_ARG(T) lhs, NBL_CONST_REF_ARG(T) rhs)
-	{
-		static array_get<T, scalar_type> getter;
-		scalar_type retval = getter(lhs, 0) * getter(rhs, 0);
-
-		static const uint32_t ArrayDim = vector_traits<T>::Dimension;
-		for (uint32_t i = 1; i < ArrayDim; ++i)
-			retval = retval + getter(lhs, i) * getter(rhs, i);
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct dot_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct cross_helper;
+template<typename T, typename U NBL_STRUCT_CONSTRAINABLE>
+struct clamp_helper;
+template<typename Integer NBL_STRUCT_CONSTRAINABLE>
+struct find_msb_helper;
+template<typename Integer NBL_STRUCT_CONSTRAINABLE>
+struct find_lsb_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct bitReverse_helper;
+template<typename Matrix NBL_STRUCT_CONSTRAINABLE>
+struct transpose_helper;
+template<typename Vector NBL_STRUCT_CONSTRAINABLE>
+struct length_helper;
+template<typename Vector NBL_STRUCT_CONSTRAINABLE>
+struct normalize_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct max_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct min_helper;
+template<typename Integer NBL_STRUCT_CONSTRAINABLE>
+struct bitCount_helper;
+template<typename LhsT, typename RhsT NBL_STRUCT_CONSTRAINABLE>
+struct mul_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct determinant_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct inverse_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct rsqrt_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct all_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct any_helper;
+template<typename T, uint16_t Bits NBL_STRUCT_CONSTRAINABLE>
+struct bitReverseAs_helper;
```

Comment:
can safely get  rid of this one

Reply:
got rid of the other one https://github.com/Devsh-Graphics-Programming/Nabla/blob/new_tgmath/include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl#L337-L350

Reply:
the code you linked has nothing to do with bitReverseAs probably because you didin't use a pemalink and just a branch link